# Evolución y equivalencia temporal de las secciones electorales

Las secciones electorales no son constantes en el tiempo. A medida que varía la distribución de la población, las secciones hacen también lo propio. La definición geométrica de éstas se revisa todos los años, pudiendose crear nuevas secciones o fusionarse otras.

Es por ello que las secciones de una elección no son las mismas que en otra, por lo que es preciso definir la forma de poder relacionar unas secciones en una elección otra.

La forma que hemos elegido es la de proximidad geográfica, es decir, si tomamos una sección X en una elección, su equivalencia en otra elección será la sección más cercana geográficamente a X. Impondremos que las secciones equivalentes pertenezcan al mismo municipio, y si el municipio ya no existe o se ha creado de la nada (cosa que pasa), haremos que se conserve la provincia.

Naturalmente, deberemos utilizar la definición geométrica que no proporcionan los ficheros shapely del INE, que leeremos con Geopandas.

Comenzamos pues cargando las librerias necesarias, incluyendo todas las que vienen con Geopandas.

In [ ]:
#dejamos este notebook como ejemplo para mostrar el proceso de ETL llevado a cabo sobre los datos iniciales del INE pero no actualizamos los archivos resultado en S3 ya que, como explicamos más abajo, no son del todo válidos

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
%%time 

# Important library for many geopython libraries
!apt install gdal-bin python-gdal python3-gdal 
# Install rtree - Geopandas requirment
!apt install python3-rtree 
# Install Geopandas
!pip install git+git://github.com/geopandas/geopandas.git
# Install descartes - Geopandas requirment
!pip install descartes 
# Install Folium for Geographic data visualization
!pip install folium
# Install plotlyExpress
!pip install plotly_express

Reading package lists... Done
Building dependency tree       
Reading state information... Done
gdal-bin is already the newest version (2.2.3+dfsg-2).
python-gdal is already the newest version (2.2.3+dfsg-2).
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  python3-numpy
Suggested packages:
  python-numpy-doc python3-nose python3-numpy-dbg
The following NEW packages will be installed:
  python3-gdal python3-numpy
0 upgraded, 2 newly installed, 0 to remove and 40 not upgraded.
Need to get 2,288 kB of archives.
After this operation, 13.2 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/main amd64 python3-numpy amd64 1:1.13.3-2ubuntu1 [1,943 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python3-gdal amd64 2.2.3+dfsg-2 [346 kB]
Fetched 2,288 kB in 1s (1,949 kB/s)
Selecting previously unselected

In [ ]:
import pandas as pd
import numpy as np
import geopandas as gpd
from shapely.geometry import Point
import matplotlib
import matplotlib.pyplot as plt 
import folium
import plotly_express as px

Ahora cargamos el fichero shapely de las elecciones de 2019, que es el mismo para las de abril y noviembre.

In [ ]:
secciones_A19 = gpd.read_file('SECC_CE_20190101.shp')

In [ ]:
secciones_A19

OBJECTID  ...                                           geometry
0             1  ...  MULTIPOLYGON (((556453.835 4752758.332, 556460...
1             2  ...  POLYGON ((502035.230 4771813.197, 502048.071 4...
2             3  ...  MULTIPOLYGON (((538984.636 4718139.608, 538985...
3             4  ...  POLYGON ((537063.531 4703664.589, 536887.844 4...
4             6  ...  POLYGON ((551570.951 4739269.962, 551570.889 4...
...         ...  ...                                                ...
36312     10673  ...  POLYGON ((17713.727 4798559.174, 17794.388 479...
36313     10659  ...  POLYGON ((75420.668 4796270.576, 75442.668 479...
36314     10189  ...  POLYGON ((59718.470 4802760.627, 59735.222 480...
36315     13100  ...  POLYGON ((151385.134 4180084.902, 151402.134 4...
36316     35989  ...  POLYGON ((372352.294 4067015.716, 372339.457 4...

[36317 rows x 24 columns]

El primer paso que damos es el crear el código completo de la sección, que incluya la elección. 

Aquí cometimos un error, confiando en que el INE e Interior hubiesen adoptado el mismo cógido para las CCAA, cosa que no es así. Nos dimos cuenta tarde, y tuviimos que arreglarlo con otro cuaderno. Las variaciones no fueron importantes, con lo que lo que hacemos en este cuaderno es válido.

In [ ]:
secciones_A19['cod_sec'] = '022019041' + secciones_A19['CCA'] + secciones_A19['CUSEC']

In [ ]:
secciones_A19['CCA'].value_counts()

01    5981
09    5071
13    4417
07    3543
10    3472
12    2169
08    1948
16    1711
02    1450
05    1381
14    1225
11     965
03     850
04     662
15     562
06     467
17     343
18      56
19      44
Name: CCA, dtype: int64

In [ ]:
secciones_A19.head()

OBJECTID  ...                cod_sec
0         1  ...  022019041160100901001
1         2  ...  022019041160101001002
2         3  ...  022019041160103101001
3         4  ...  022019041160103301001
4         6  ...  022019041160103701001

[5 rows x 25 columns]

In [ ]:
len(secciones_A19['cod_sec'].unique())

36317

Cargamos el dataset de las elecciones de abril de 2019.

In [ ]:
strings = {'Sección' : 'str', 'cod_ccaa' : 'str', 'cod_prov' : 'str', 'cod_mun' : 'str', 'cod_sec' : 'str'}

In [ ]:
df_eleccion_A19 = pd.read_csv('gen_A19_unif_cols_prov.txt', dtype = strings)

In [ ]:
df_eleccion_A19

Sección  ...                                       dict_res_ord
0      022019041010400101001  ...  [('PSOE', 326), ('PP', 149), ('Cs', 131), ('VO...
1      022019041010400201001  ...  [('PSOE', 380), ('PP', 127), ('VOX', 113), ('C...
2      022019041010400301001  ...  [('PP', 162), ('PSOE', 131), ('VOX', 103), ('C...
3      022019041010400301002  ...  [('PSOE', 241), ('PP', 239), ('VOX', 218), ('C...
4      022019041010400301003  ...  [('VOX', 282), ('PP', 274), ('PSOE', 252), ('C...
...                      ...  ...                                                ...
36312  022019041195200108011  ...  [('Otros', 288), ('PP', 250), ('PSOE', 194), (...
36313  022019041195200108012  ...  [('PP', 421), ('PSOE', 238), ('VOX', 212), ('C...
36314  022019041195200108013  ...  [('PP', 175), ('PSOE', 168), ('VOX', 144), ('C...
36315  022019041195200108014  ...  [('PP', 222), ('VOX', 103), ('PSOE', 98), ('Ot...
36316  022019041195200108015  ...  [('PP', 300), ('PSOE', 267), ('VOX', 203), ('C...

[36317 rows x 97 columns]

Aquí lo que comprobamos es que los códigos de las CCAA son distintos, según el INE e Interior. Lo rectificamos en otro cuaderno.

Con todo, estas filas que siguen son posteriores a redactar este cuaderno, luego lo que hicimos al principio es continuar creando los cógidos completos de las secciones en los demás ficheros shapely.

In [ ]:
df_eleccion_A19['parte'] = df_eleccion_A19['Sección'].str[0:9]

In [ ]:
df_eleccion_A19['parte'].value_counts()

022019041    36317
Name: parte, dtype: int64

In [ ]:
secciones_A19['cod_sec'].str[0:9].value_counts()

022019041    36317
Name: cod_sec, dtype: int64

In [ ]:
df_eleccion_A19['cod_ccaa'].value_counts()

01    5981
09    5071
12    4417
08    3543
17    3472
11    2169
07    1948
14    1711
02    1450
05    1381
15    1225
10     965
03     850
04     662
13     562
06     467
16     343
18      56
19      44
Name: cod_ccaa, dtype: int64

In [ ]:
secciones_A19['CCA'].value_counts()

01    5981
09    5071
13    4417
07    3543
10    3472
12    2169
08    1948
16    1711
02    1450
05    1381
14    1225
11     965
03     850
04     662
15     562
06     467
17     343
18      56
19      44
Name: CCA, dtype: int64

Así, ahora cargamos el fichero shapey de noviembre de 2019, y creamos el código de sección, que está mal, pero entonces no los sabíamos.

In [ ]:
secciones_N19 = gpd.read_file('SECC_CE_20190101.shp')

In [ ]:
secciones_N19['cod_sec'] = '022019111' + secciones_N19['CCA'] + secciones_N19['CUSEC']

In [ ]:
secciones_N19

OBJECTID  ...                cod_sec
0             1  ...  022019111160100901001
1             2  ...  022019111160101001002
2             3  ...  022019111160103101001
3             4  ...  022019111160103301001
4             6  ...  022019111160103701001
...         ...  ...                    ...
36312     10673  ...  022019111121509302002
36313     10659  ...  022019111121590203001
36314     10189  ...  022019111121503103003
36315     13100  ...  022019111012101701002
36316     35989  ...  022019111012906707064

[36317 rows x 25 columns]

Hacemos lo mismo con el de 2015.

In [ ]:
secciones_D15 = gpd.read_file('SECC_CE_20150101_01_R_INE.shp')

In [ ]:
secciones_D15['cod_sec'] = '022015121' + secciones_D15['CCA'] + secciones_D15['CUSEC']

In [ ]:
secciones_D15

CUSEC  ...                cod_sec
0      0100101001  ...  022015121160100101001
1      0100101002  ...  022015121160100101002
2      0100201001  ...  022015121160100201001
3      0100201002  ...  022015121160100201002
4      0100201003  ...  022015121160100201003
...           ...  ...                    ...
36224  5200108011  ...  022015121195200108011
36225  5200108012  ...  022015121195200108012
36226  5200108013  ...  022015121195200108013
36227  5200108014  ...  022015121195200108014
36228  5200108015  ...  022015121195200108015

[36229 rows x 20 columns]

In [ ]:
secciones_D15['CCA'].value_counts()

01    5926
09    5050
13    4341
07    3553
10    3473
12    2196
08    1952
16    1735
02    1448
05    1367
14    1232
11     966
03     871
04     653
15     562
06     468
17     338
18      54
19      44
Name: CCA, dtype: int64

Y con el de 2016.

In [ ]:
secciones_J16 = gpd.read_file('SECC_CE_20160101.shp')

In [ ]:
secciones_J16['cod_sec'] = '022016061' + secciones_J16['CCA'] + secciones_J16['CUSEC']

In [ ]:
secciones_J16

CUSEC  ...                cod_sec
0      0100101001  ...  022016061160100101001
1      0100101002  ...  022016061160100101002
2      0100201001  ...  022016061160100201001
3      0100201002  ...  022016061160100201002
4      0100201003  ...  022016061160100201003
...           ...  ...                    ...
36210  5200108011  ...  022016061195200108011
36211  5200108012  ...  022016061195200108012
36212  5200108013  ...  022016061195200108013
36213  5200108014  ...  022016061195200108014
36214  5200108015  ...  022016061195200108015

[36215 rows x 23 columns]

In [ ]:
secciones_J16['CCA'].value_counts()

01    5927
09    5046
13    4341
07    3556
10    3470
12    2187
08    1953
16    1735
02    1448
05    1367
14    1232
11     966
03     868
04     653
15     562
06     468
17     338
18      54
19      44
Name: CCA, dtype: int64

Y finalmente con el de 2011.

In [ ]:
secciones_N11 = gpd.read_file('SECC_CE_20110101_03_R_INE.shp')

In [ ]:
secciones_N11['cod_sec'] = '022011111' + secciones_N11['CCA'] + secciones_N11['CUSEC']

In [ ]:
secciones_N11

OBJECTID  ...                cod_sec
0           1.0  ...  022011111160100101001
1           2.0  ...  022011111160100101002
2           3.0  ...  022011111160100201001
3           4.0  ...  022011111160100201002
4           5.0  ...  022011111160100201003
...         ...  ...                    ...
35956   35958.0  ...  022011111195200108010
35957   35959.0  ...  022011111195200108011
35958   35960.0  ...  022011111195200108012
35959   35961.0  ...  022011111195200108013
35960   35962.0  ...  022011111195200108014

[35961 rows x 24 columns]

In [ ]:
secciones_N11['CCA'].value_counts()

01    5796
09    5019
13    4271
07    3538
10    3478
12    2249
08    1942
16    1739
02    1451
05    1330
14    1220
11     969
03     878
04     630
15     558
06     463
17     336
18      53
19      41
Name: CCA, dtype: int64

Bien, como habrá que buscar las secciones equivalentes a cada una, y en teoría todas son distintas, lo que hacemos en primer lugar es fusionar los datasets de las cinco shapelys.

In [ ]:
df_unif = secciones_A19.append(secciones_N19, ignore_index=True)

In [ ]:
df_unif = df_unif.append(secciones_J16, ignore_index=True)

In [ ]:
df_unif = df_unif.append(secciones_D15, ignore_index=True)

In [ ]:
df_unif = df_unif.append(secciones_N11, ignore_index=True)

In [ ]:
df_unif

OBJECTID       CUSEC  ...   Shape_Le_1 Shape_Area
0            1.0  0100901001  ...          NaN        NaN
1            2.0  0101001002  ...          NaN        NaN
2            3.0  0103101001  ...          NaN        NaN
3            4.0  0103301001  ...          NaN        NaN
4            6.0  0103701001  ...          NaN        NaN
...          ...         ...  ...          ...        ...
181034   35958.0  5200108010  ...   868.395859        NaN
181035   35959.0  5200108011  ...  7347.338748        NaN
181036   35960.0  5200108012  ...  1571.615853        NaN
181037   35961.0  5200108013  ...  4691.060767        NaN
181038   35962.0  5200108014  ...   790.887241        NaN

[181039 rows x 27 columns]

Comprobamos que seguimos estando en un fichero shapely preguntado por la proyección.

In [ ]:
df_unif.crs

<Projected CRS: EPSG:25830>
Name: ETRS89 / UTM zone 30N
Axis Info [cartesian]:
- E[east]: Easting (metre)
- N[north]: Northing (metre)
Area of Use:
- name: Europe between 6°W and 0°W: Faroe Islands offshore; Ireland - offshore; Jan Mayen - offshore; Norway including Svalbard - offshore; Spain - onshore and offshore.
- bounds: (-5.9999999999999, 35.265663028, 1.7053025658242e-13, 80.489344496333)
Coordinate Operation:
- name: UTM zone 30N
- method: Transverse Mercator
Datum: European Terrestrial Reference System 1989 ensemble
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

Hay muchos metadatos en los shapely que no nos serán útiles. Nos quedamos con los códigos de secciones y su definición geométrica.

In [ ]:
df_unif.columns

Index(['OBJECTID', 'CUSEC', 'CUMUN', 'CSEC', 'CDIS', 'CMUN', 'CPRO', 'CCA',
       'CUDIS', 'CLAU2', 'NPRO', 'NCA', 'CNUT0', 'CNUT1', 'CNUT2', 'CNUT3',
       'ESTADO', 'OBS', 'Shape_Leng', 'Shape_area', 'Shape_len', 'SUPERF_M2',
       'NMUN', 'geometry', 'cod_sec', 'Shape_Le_1', 'Shape_Area'],
      dtype='object')

In [ ]:
df_unif = df_unif[['cod_sec', 'CUSEC', 'CUMUN', 'CPRO', 'geometry']]

In [ ]:
df_unif

cod_sec  ...                                           geometry
0       022019041160100901001  ...  MULTIPOLYGON (((556453.835 4752758.332, 556460...
1       022019041160101001002  ...  POLYGON ((502035.230 4771813.197, 502048.071 4...
2       022019041160103101001  ...  MULTIPOLYGON (((538984.636 4718139.608, 538985...
3       022019041160103301001  ...  POLYGON ((537063.531 4703664.589, 536887.844 4...
4       022019041160103701001  ...  POLYGON ((551570.951 4739269.962, 551570.889 4...
...                       ...  ...                                                ...
181034  022011111195200108010  ...  POLYGON ((504318.195 3904957.402, 504318.033 3...
181035  022011111195200108011  ...  POLYGON ((503855.015 3905844.061, 503855.334 3...
181036  022011111195200108012  ...  POLYGON ((505731.785 3904249.139, 505637.515 3...
181037  022011111195200108013  ...  POLYGON ((506068.859 3903736.075, 506069.536 3...
181038  022011111195200108014  ...  POLYGON ((504042.428 3905477.470, 504038.593 3...

[181039 rows x 5 columns]

In [ ]:
df_unif.dtypes

cod_sec       object
CUSEC         object
CUMUN         object
CPRO          object
geometry    geometry
dtype: object

Ahora bien, para calcular la posición de una sección, la forma más cómoda es definir, o más bien calcular, su centroide. Geopandas tiene un método dedicado a ello. Creamos una nueva columna con este dato, que es simplememte un punto.

In [ ]:
from shapely.geometry import Polygon, LineString, Point

In [ ]:
df_unif['Centroide'] = df_unif['geometry'].centroid

/usr/local/lib/python3.7/dist-packages/geopandas/geodataframe.py:1345: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df_unif

cod_sec  ...                       Centroide
0       022019041160100901001  ...  POINT (555853.735 4749228.022)
1       022019041160101001002  ...  POINT (500019.424 4771906.777)
2       022019041160103101001  ...  POINT (535052.148 4710767.820)
3       022019041160103301001  ...  POINT (535297.829 4704792.205)
4       022019041160103701001  ...  POINT (545408.849 4733097.078)
...                       ...  ...                             ...
181034  022011111195200108010  ...  POINT (504226.917 3904997.784)
181035  022011111195200108011  ...  POINT (503591.236 3905281.004)
181036  022011111195200108012  ...  POINT (505545.192 3904295.175)
181037  022011111195200108013  ...  POINT (506046.231 3903459.669)
181038  022011111195200108014  ...  POINT (503857.192 3905480.595)

[181039 rows x 6 columns]

Por comodidad creamos otra columna con el código que identifica la elección, siendo 02 una elección general, seguida del año y del mes.

In [ ]:
df_unif['Elección'] = df_unif['cod_sec'].str[0:8] 

/usr/local/lib/python3.7/dist-packages/geopandas/geodataframe.py:1345: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df_unif

cod_sec  ...  Elección
0       022019041160100901001  ...  02201904
1       022019041160101001002  ...  02201904
2       022019041160103101001  ...  02201904
3       022019041160103301001  ...  02201904
4       022019041160103701001  ...  02201904
...                       ...  ...       ...
181034  022011111195200108010  ...  02201111
181035  022011111195200108011  ...  02201111
181036  022011111195200108012  ...  02201111
181037  022011111195200108013  ...  02201111
181038  022011111195200108014  ...  02201111

[181039 rows x 7 columns]

Como ya vimos, la equivalencia de una sección X de una elección, porgamos 2015 en otra elección, 2016, será la sección de esta última elección que tenga un centroide más cercano al de X, siempre que esté en su mismo municipio o provincia.

Para calcular distancias Geopandas cuenta también con un método .distance().

Podemos ver un ejemplo de la distancia entre los centrioides de dos secciones, en este caso unos 520 kilómetros.

In [ ]:
a = df_unif['Centroide'][586]

In [ ]:
b = df_unif['Centroide'][10587]

In [ ]:
a.distance(b)

520189.72612155194

Tomamos ahora la sección de índice 10587, que es una de las elecciones de abril de 2019. Resulta ser del municipio de código 17062.

In [ ]:
secc = df_unif['cod_sec'][10587]
secc

'022019041091706201003'

In [ ]:
cent = df_unif['Centroide'][10587]


In [ ]:
mun = df_unif['CUMUN'][10587]
mun

'17062'

Ahora queremos buscar su equivalente en las elecciones de 2011. Buscamos las secciones de ese municipio en esa elección, y obtenemos un df, df_aux.

In [ ]:
elect = '02201111'

In [ ]:
df_aux = df_unif.loc[df_unif['Elección'] == elect].loc[df_unif['CUMUN'] == mun]

In [ ]:
df_aux

cod_sec  ...  Elección
157092  022011111091706201001  ...  02201111
157093  022011111091706201002  ...  02201111
157094  022011111091706201003  ...  02201111
157095  022011111091706201004  ...  02201111
157096  022011111091706201005  ...  02201111

[5 rows x 7 columns]

La sección equivalente será aquella cuyo centroide tenga la menor distancia con el centroide de la sección 10587, que es desde luego la de índice 157094.

In [ ]:
df_aux['dist'] = df_aux['Centroide'].apply(lambda x : x.distance(cent))

In [ ]:
df_aux

cod_sec       CUSEC  ...  Elección         dist
157092  022011111091706201001  1706201001  ...  02201111  2102.533307
157093  022011111091706201002  1706201002  ...  02201111  2733.288140
157094  022011111091706201003  1706201003  ...  02201111     0.541192
157095  022011111091706201004  1706201004  ...  02201111  2474.517126
157096  022011111091706201005  1706201005  ...  02201111  4142.634792

[5 rows x 8 columns]

In [ ]:
ind = df_aux['dist'].idxmin()
ind

157094

In [ ]:
sec = df_aux['cod_sec'][ind]
sec

'022011111091706201003'

Se trata de iterar este proceso que acabamos de ver a lo largo de todas las secciones de las cinco elecciones, unas 180 mil. Desde luego, la sección equivalente a una sección en su misma elección es ella misma.

Definimos una columna para almecenar las secciones más cercanas en una determinada elección.

In [ ]:
df_unif['cercana N11'] = ''
df_unif['cercana D15'] = ''
df_unif['cercana J16'] = ''
df_unif['cercana A19'] = ''
df_unif['cercana N19'] = ''

/usr/local/lib/python3.7/dist-packages/geopandas/geodataframe.py:1345: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df_unif

cod_sec       CUSEC  ... cercana A19 cercana N19
0       022019041160100901001  0100901001  ...                        
1       022019041160101001002  0101001002  ...                        
2       022019041160103101001  0103101001  ...                        
3       022019041160103301001  0103301001  ...                        
4       022019041160103701001  0103701001  ...                        
...                       ...         ...  ...         ...         ...
181034  022011111195200108010  5200108010  ...                        
181035  022011111195200108011  5200108011  ...                        
181036  022011111195200108012  5200108012  ...                        
181037  022011111195200108013  5200108013  ...                        
181038  022011111195200108014  5200108014  ...                        

[181039 rows x 12 columns]

In [ ]:
secciones_N19.dtypes

OBJECTID         int64
CUSEC           object
CUMUN           object
CSEC            object
CDIS            object
CMUN            object
CPRO            object
CCA             object
CUDIS           object
CLAU2           object
NPRO            object
NCA             object
CNUT0           object
CNUT1           object
CNUT2           object
CNUT3           object
ESTADO          object
OBS             object
Shape_Leng     float64
Shape_area     float64
Shape_len      float64
SUPERF_M2        int64
NMUN            object
geometry      geometry
cod_sec         object
dtype: object

Tenemos que tener en cuenta que los municipios no son los mismos en cada elección, por lo que tenemos que saber cuales cambian, y aplicar en criterio de misma provincia.

En las celdas que siguen buscamos qué municipios son lo que no están en las cinco elecciones, y obtenemos sus códigos, algunos se han creado y otros han desaparecido; en total son 23. Hallamos sus códigos mediante una manipulación de sets.

In [ ]:
N19 = set(secciones_N19['CUMUN'])

In [ ]:
len(N19)

8131

In [ ]:
N19

{'37128',
 '07039',
 '47169',
 '04053',
 '21060',
 '16094',
 '22217',
 '05069',
 '12060',
 '24175',
 '34010',
 '23069',
 '42115',
 '10142',
 '04022',
 '19267',
 '47060',
 '42145',
 '25093',
 '26052',
 '13062',
 '35033',
 '28030',
 '08021',
 '16250',
 '13029',
 '37165',
 '46183',
 '16145',
 '10073',
 '49267',
 '14055',
 '17018',
 '10027',
 '09277',
 '10082',
 '04023',
 '18069',
 '16177',
 '40204',
 '03055',
 '19306',
 '37302',
 '25132',
 '50175',
 '44017',
 '37252',
 '04071',
 '31065',
 '26003',
 '49171',
 '15012',
 '07060',
 '48036',
 '12138',
 '37251',
 '39079',
 '50082',
 '11039',
 '49031',
 '09072',
 '26088',
 '31194',
 '16265',
 '18111',
 '18035',
 '24021',
 '14062',
 '31076',
 '08120',
 '10034',
 '33035',
 '18056',
 '35019',
 '40054',
 '50205',
 '44241',
 '12117',
 '20007',
 '10018',
 '08014',
 '09414',
 '10011',
 '05154',
 '40219',
 '22042',
 '10214',
 '43124',
 '41047',
 '03119',
 '09424',
 '02078',
 '10196',
 '40107',
 '13073',
 '22908',
 '29008',
 '25247',
 '37361',
 '09329',


In [ ]:
N11 = set(secciones_N11['CUMUN'])

In [ ]:
len(N11)

8116

In [ ]:
dif = N19.difference(N11)

In [ ]:
len(dif)

19

In [ ]:
dif

{'04904',
 '06903',
 '10904',
 '10905',
 '11903',
 '14901',
 '14902',
 '15902',
 '18065',
 '18077',
 '18106',
 '18914',
 '18915',
 '18916',
 '21902',
 '29903',
 '29904',
 '36902',
 '41904'}

In [ ]:
dif1 = N11.difference(N19)

In [ ]:
dif1

{'15026', '15063', '36011', '36012'}

In [ ]:
'04904' not in dif

False

In [ ]:
dif = dif.union(dif1)

In [ ]:
len(dif)

23

Ahora lo que hacemos es crear un objeto groupby mediante el código de municipio y la elección, que hará mucho más eficiente el proceso de encontrar las secciones equivalentes.

Como habrá unos pocos casos que tendremos que usar la provincia, creamos un objeto gruopby análogo con las provincias.

In [ ]:
grouped = df_unif.groupby(["CUMUN", "Elección"])

In [ ]:
grouped.groups

{('01001', '02201111'): [145078, 145079], ('01001', '02201512'): [108849, 108850], ('01001', '02201606'): [72634, 72635], ('01001', '02201904'): [32923, 32961], ('01001', '02201911'): [69240, 69278], ('01002', '02201111'): [145080, 145081, 145082, 145083, 145084, 145085, 145086], ('01002', '02201512'): [108851, 108852, 108853, 108854, 108855, 108856, 108857], ('01002', '02201606'): [72636, 72637, 72638, 72639, 72640, 72641, 72642], ('01002', '02201904'): [11, 13, 20, 21, 38, 39, 48], ('01002', '02201911'): [36328, 36330, 36337, 36338, 36355, 36356, 36365], ('01003', '02201111'): [145087], ('01003', '02201512'): [108858], ('01003', '02201606'): [72643], ('01003', '02201904'): [49], ('01003', '02201911'): [36366], ('01004', '02201111'): [145088], ('01004', '02201512'): [108859], ('01004', '02201606'): [72644], ('01004', '02201904'): [12], ('01004', '02201911'): [36329], ('01006', '02201111'): [145089], ('01006', '02201512'): [108860], ('01006', '02201606'): [72645], ('01006', '02201904')

In [ ]:
grouped['cod_sec']

In [ ]:
a = grouped['Centroide']

In [ ]:
grouped.get_group(('01001', '02201111'))

cod_sec       CUSEC  ... cercana A19 cercana N19
145078  022011111160100101001  0100101001  ...                        
145079  022011111160100101002  0100101002  ...                        

[2 rows x 12 columns]

In [ ]:
grouped_prov = df_unif.groupby(["CPRO", "Elección"])

In [ ]:
grouped_prov.groups

{('01', '02201111'): [145078, 145079, 145080, 145081, 145082, 145083, 145084, 145085, 145086, 145087, 145088, 145089, 145090, 145091, 145092, 145093, 145094, 145095, 145096, 145097, 145098, 145099, 145100, 145101, 145102, 145103, 145104, 145105, 145106, 145107, 145108, 145109, 145110, 145111, 145112, 145113, 145114, 145115, 145116, 145117, 145118, 145119, 145120, 145121, 145122, 145123, 145124, 145125, 145126, 145127, 145128, 145129, 145130, 145131, 145132, 145133, 145134, 145135, 145136, 145137, 145138, 145139, 145140, 145141, 145142, 145143, 145144, 145145, 145146, 145147, 145148, 145149, 145150, 145151, 145152, 145153, 145154, 145155, 145156, 145157, 145158, 145159, 145160, 145161, 145162, 145163, 145164, 145165, 145166, 145167, 145168, 145169, 145170, 145171, 145172, 145173, 145174, 145175, 145176, 145177, ...], ('01', '02201512'): [108849, 108850, 108851, 108852, 108853, 108854, 108855, 108856, 108857, 108858, 108859, 108860, 108861, 108862, 108863, 108864, 108865, 108866, 108867,

A continuación definimos la función que nos dará las secciones equivalentes para una determinada elección (elect), y nos la asiganará en la columna (col).

Lo que hace es iterar en cada sección de las 180 mil. Primero selecciona un dataset auxiliar con el mismo municipio de la sección en la eleccin elect. En el caso que el municipio sea de los 23 que ha cambiado tomará su provincia entera.

Después calcula la distancia entre centroides, quedándose con la mínima, e identificando la sección correspondiente mediante el índice, y finalmente rellena el código de la sección en la columna col.

Es en definitiva la iteración del proceso que vimos anteriormente.

In [ ]:
def cercanos(elect, col):

  for ind in range(len(df_unif)):
    cent = df_unif['Centroide'][ind]
    mun = df_unif['CUMUN'][ind]
    prov = df_unif['CPRO'][ind]
  
    if mun not in dif:
      df_aux = grouped.get_group((mun, elect))

    else:
      df_aux = grouped_prov.get_group((prov, elect))

    df_aux['dist'] = df_aux['Centroide'].apply(lambda x : x.distance(cent))
    idx = df_aux['dist'].idxmin()
    sec = df_aux['cod_sec'][idx]

    df_unif[col][ind] = sec

    if ind % 2500 == 0:
      print(ind)

  
  

Se trata de un proceso lento, que para cada elección tarda unos 50 minutos. La primera vez lo ejecutamos solo para la elección de 2011.

In [ ]:
cercanos(elect = '02201111', col = 'cercana N11')

0
2500
5000
7500
10000
12500
15000
17500
20000
22500
25000
27500
30000
32500
35000
37500
40000
42500
45000
47500
50000
52500
55000
57500
60000
62500
65000
67500
70000
72500
75000
77500
80000
82500
85000
87500
90000
92500
95000
97500
100000
102500
105000
107500
110000
112500
115000
117500
120000
122500
125000
127500
130000
132500
135000
137500
140000
142500
145000
147500
150000
152500
155000
157500
160000
162500
165000
167500
170000
172500
175000
177500
180000


In [ ]:
df_unif.to_csv('similitud_secciones_1.csv', sep = ',', index = False)

En la siguiente sesión la ejecutamos para las otras cuatro elecciones, comenzando con la de 2015...

In [ ]:
cercanos(elect = '02201512', col = 'cercana D15')

0
2500
5000
7500
10000
12500
15000
17500
20000
22500
25000
27500
30000
32500
35000
37500
40000
42500
45000
47500
50000
52500
55000
57500
60000
62500
65000
67500
70000
72500
75000
77500
80000
82500
85000
87500
90000
92500
95000
97500
100000
102500
105000
107500
110000
112500
115000
117500
120000
122500
125000
127500
130000
132500
135000
137500
140000
142500
145000
147500
150000
152500
155000
157500
160000
162500
165000
167500
170000
172500
175000
177500
180000


Después la de 2016...

In [ ]:
cercanos(elect = '02201606', col = 'cercana J16')

0
2500
5000
7500
10000
12500
15000
17500
20000
22500
25000
27500
30000
32500
35000
37500
40000
42500
45000
47500
50000
52500
55000
57500
60000
62500
65000
67500
70000
72500
75000
77500
80000
82500
85000
87500
90000
92500
95000
97500
100000
102500
105000
107500
110000
112500
115000
117500
120000
122500
125000
127500
130000
132500
135000
137500
140000
142500
145000
147500
150000
152500
155000
157500
160000
162500
165000
167500
170000
172500
175000
177500
180000


...abril de 2019...

In [ ]:
cercanos(elect = '02201904', col = 'cercana A19')

0
2500
5000
7500
10000
12500
15000
17500
20000
22500
25000
27500
30000
32500
35000
37500
40000
42500
45000
47500
50000
52500
55000
57500
60000
62500
65000
67500
70000
72500
75000
77500
80000
82500
85000
87500
90000
92500
95000
97500
100000
102500
105000
107500
110000
112500
115000
117500
120000
122500
125000
127500
130000
132500
135000
137500
140000
142500
145000
147500
150000
152500
155000
157500
160000
162500
165000
167500
170000
172500
175000
177500
180000


... y noviembre de 2019.

In [ ]:
cercanos(elect = '02201911', col = 'cercana N19')

0
2500
5000
7500
10000
12500
15000
17500
20000
22500
25000
27500
30000
32500
35000
37500
40000
42500
45000
47500
50000
52500
55000
57500
60000
62500
65000
67500
70000
72500
75000
77500
80000
82500
85000
87500
90000
92500
95000
97500
100000
102500
105000
107500
110000
112500
115000
117500
120000
122500
125000
127500
130000
132500
135000
137500
140000
142500
145000
147500
150000
152500
155000
157500
160000
162500
165000
167500
170000
172500
175000
177500
180000


Este fue el resultado de la segunda sesión, en la que faltan las secciones equivalentes en las elecciones de 2011, que ya habíamos calculado.

In [ ]:
df_unif

cod_sec  ...            cercana N19
0       022019041160100901001  ...  022019111160100901001
1       022019041160101001002  ...  022019111160101001002
2       022019041160103101001  ...  022019111160103101001
3       022019041160103301001  ...  022019111160103301001
4       022019041160103701001  ...  022019111160103701001
...                       ...  ...                    ...
181034  022011111195200108010  ...  022019111195200108010
181035  022011111195200108011  ...  022019111195200108011
181036  022011111195200108012  ...  022019111195200108012
181037  022011111195200108013  ...  022019111195200108013
181038  022011111195200108014  ...  022019111195200108014

[181039 rows x 12 columns]

In [ ]:
df_unif['Elección'].unique()

array(['02201904', '02201911', '02201606', '02201512', '02201111'],
      dtype=object)

Realmente, ya no nos interesan las columnas de geometry o centroide, solo las secciones, y con esas columnas nos quedamos.

In [ ]:
df_unit_simp = df_unif[['cod_sec', 'CUSEC', 'CUMUN', 'CPRO', 'Elección', 'cercana N11', 'cercana D15', 'cercana J16', 'cercana A19', 'cercana N19']]

In [ ]:
df_unit_simp

cod_sec  ...            cercana N19
0       022019041160100901001  ...  022019111160100901001
1       022019041160101001002  ...  022019111160101001002
2       022019041160103101001  ...  022019111160103101001
3       022019041160103301001  ...  022019111160103301001
4       022019041160103701001  ...  022019111160103701001
...                       ...  ...                    ...
181034  022011111195200108010  ...  022019111195200108010
181035  022011111195200108011  ...  022019111195200108011
181036  022011111195200108012  ...  022019111195200108012
181037  022011111195200108013  ...  022019111195200108013
181038  022011111195200108014  ...  022019111195200108014

[181039 rows x 10 columns]

In [ ]:
df_unit_simp.to_csv('similitud_secciones_1.csv', sep = ',', index = False)

Faltan las seccines de 2011, que cargamos de su fichero.

In [ ]:
cols_str = {'cercana N11' : 'str'}

In [ ]:
unif_prov = pd.read_csv('similitud_secciones.csv', dtype = cols_str)

In [ ]:
unif_prov

cod_sec       CUSEC  ...  cercana A19  cercana N19
0       022019041160100901001   100901001  ...          NaN          NaN
1       022019041160101001002   101001002  ...          NaN          NaN
2       022019041160103101001   103101001  ...          NaN          NaN
3       022019041160103301001   103301001  ...          NaN          NaN
4       022019041160103701001   103701001  ...          NaN          NaN
...                       ...         ...  ...          ...          ...
181034  022011111195200108010  5200108010  ...          NaN          NaN
181035  022011111195200108011  5200108011  ...          NaN          NaN
181036  022011111195200108012  5200108012  ...          NaN          NaN
181037  022011111195200108013  5200108013  ...          NaN          NaN
181038  022011111195200108014  5200108014  ...          NaN          NaN

[181039 rows x 12 columns]

In [ ]:
df_unit_simp

cod_sec  ...            cercana N19
0       022019041160100901001  ...  022019111160100901001
1       022019041160101001002  ...  022019111160101001002
2       022019041160103101001  ...  022019111160103101001
3       022019041160103301001  ...  022019111160103301001
4       022019041160103701001  ...  022019111160103701001
...                       ...  ...                    ...
181034  022011111195200108010  ...  022019111195200108010
181035  022011111195200108011  ...  022019111195200108011
181036  022011111195200108012  ...  022019111195200108012
181037  022011111195200108013  ...  022019111195200108013
181038  022011111195200108014  ...  022019111195200108014

[181039 rows x 10 columns]

Copiamos la columna de 2011, simplemente.

In [ ]:
df_unit_simp['cercana N11'] = unif_prov['cercana N11']

In [ ]:
df_unit_simp

cod_sec  ...            cercana N19
0       022019041160100901001  ...  022019111160100901001
1       022019041160101001002  ...  022019111160101001002
2       022019041160103101001  ...  022019111160103101001
3       022019041160103301001  ...  022019111160103301001
4       022019041160103701001  ...  022019111160103701001
...                       ...  ...                    ...
181034  022011111195200108010  ...  022019111195200108010
181035  022011111195200108011  ...  022019111195200108011
181036  022011111195200108012  ...  022019111195200108012
181037  022011111195200108013  ...  022019111195200108013
181038  022011111195200108014  ...  022019111195200108014

[181039 rows x 10 columns]

Y finalmente guardamos el fichero csv con todas las secciones. Recordamos que hubo que variar algo los códigos de las secciones debido a los distintos códigos de las ccaa entre el INE e Interior, lo cual hicimos en otro cuaderno.

In [ ]:
df_unit_simp.to_csv('similitud_secciones_def.csv', sep = ',', index = False)